phoenix-hive/src/main/java/org/apache/phoenix/hive/ql/index/IndexPredicateAnalyzer.java

```diff
@@ -0,0 +1,486 @@
+        if (allowed == null) {
+            return expr;
+        }
+
+        String udfName = genericUDF.getUdfName();
+        if (!allowed.contains(genericUDF.getUdfName())) {
+            return expr;
+        }
+
+        String[] fields = null;
+        if (extracted.length > 2) {
+            ExprNodeFieldDesc fieldDesc = (ExprNodeFieldDesc) extracted[2];
+            if (!isValidField(fieldDesc)) {
+                return expr;
+            }
+            fields = ExprNodeDescUtils.extractFields(fieldDesc);
+        }
+
+        // We also need to update the expr so that the index query can be
+        // generated.
+        // Note that, hive does not support UDFToDouble etc in the query text.
+        List<ExprNodeDesc> list = new ArrayList<ExprNodeDesc>();
+        list.add(expr1);
+        list.add(expr2);
+        expr = new ExprNodeGenericFuncDesc(expr.getTypeInfo(), expr.getGenericUDF(), list);
+
+        searchConditions.add(new IndexSearchCondition(columnDesc, udfName, constantDesc, expr, fields));
+
+        // we converted the expression to a search condition, so
+        // remove it from the residual predicate
+        return fields == null ? null : expr;
+    }
+
+    private boolean isValidField(ExprNodeFieldDesc field) {
+        return fieldValidator == null || fieldValidator.validate(field);
+    }
+
+    /**
+     * Translates search conditions back to ExprNodeDesc form (as a left-deep
+     * conjunction).
+     *
+     * @param searchConditions (typically produced by analyzePredicate)
+     * @return ExprNodeGenericFuncDesc form of search conditions
+     */
+    public ExprNodeGenericFuncDesc translateSearchConditions(List<IndexSearchCondition> searchConditions) {
+
+        ExprNodeGenericFuncDesc expr = null;
+
+        for (IndexSearchCondition searchCondition : searchConditions) {
+            if (expr == null) {
+                expr = searchCondition.getComparisonExpr();
+                continue;
+            }
+
+            List<ExprNodeDesc> children = new ArrayList<ExprNodeDesc>();
+            children.add(expr);
+            children.add(searchCondition.getComparisonExpr());
+            expr = new ExprNodeGenericFuncDesc(TypeInfoFactory.booleanTypeInfo, FunctionRegistry.getGenericUDFForAnd(), children);
+        }
+
+        return expr;
+    }
+
+    public void setAcceptsFields(boolean acceptsFields) {
+        this.acceptsFields = acceptsFields;
+    }
+
+    public static interface FieldValidator {
+        boolean validate(ExprNodeFieldDesc exprNodeDesc);
+    }
+
+    public static IndexPredicateAnalyzer createAnalyzer(boolean equalOnly) {
+        IndexPredicateAnalyzer analyzer = new IndexPredicateAnalyzer();
+        analyzer.addComparisonOp("org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqual");
+
+        if (equalOnly) {
+            return analyzer;
+        }
+
+        analyzer.addComparisonOp("org.apache.hadoop.hive.ql.udf.generic.GenericUDFOPEqualOrGreaterThan");
```

Comment:
Actually not, but it was done in the same way as it's done in Hive's IndexPredicateAnalyzer.java